src/textgen.rs

```diff
@@ -248,3 +250,86 @@ impl<T: Seek + io::Read> WordSelector for RawWordSelector<T> {
         Ok(word)
     }
 }
+
+/// Wraps another word selector, taking words from it and adding punctuation to the end of or
+/// around words with a configurable chance. Will capitalize the next word when an end-of-sentence
+/// punctuation mark is used.
+pub struct PunctuatedWordSelector {
+    selector: Box<dyn WordSelector>,
+    next_is_capital: bool,
+    punctuation_chance: f64,
+}
+
+enum PunctuationType {
+    Capitaizing(char),
+    Ending(char),
+    Surrounding(char, char),
+}
+
+const PUNCTUATION: [PunctuationType; 12] = [
+    PunctuationType::Capitaizing('!'),
+    PunctuationType::Capitaizing('?'),
+    PunctuationType::Capitaizing('.'),
+    PunctuationType::Ending(','),
+    PunctuationType::Ending(':'),
+    PunctuationType::Ending(';'),
+    PunctuationType::Surrounding('\'', '\''),
+    PunctuationType::Surrounding('"', '"'),
+    PunctuationType::Surrounding('(', ')'),
+    PunctuationType::Surrounding('{', '}'),
+    PunctuationType::Surrounding('<', '>'),
+    PunctuationType::Surrounding('[', ']'),
+];
+
+impl PunctuatedWordSelector {
+    /// Creates a PunctuatedWordSelector from another WordSelector, allowing the selection of the
+    /// chance of punctuation.
+    pub fn from_word_selector(
+        word_selector: Box<dyn WordSelector>,
+        punctuation_chance: f64,
+    ) -> Self {
+        Self {
+            selector: word_selector,
+            next_is_capital: true,
+            punctuation_chance,
+        }
+    }
+}
+
+impl WordSelector for PunctuatedWordSelector {
+    fn new_word(&mut self) -> Result<String, io::Error> {
+        let mut rng = rand::thread_rng();
+
+        let mut word = self.selector.new_word()?;
+
+        let will_punctuate = rng.gen_bool(self.punctuation_chance);
+        if will_punctuate || self.next_is_capital {
+            let mut chars: VecDeque<char> = word.chars().collect();
+            if self.next_is_capital {
+                // some unicode chars map to multiple chars when uppercased.
+                for c in chars.pop_front().expect("got empty word").to_uppercase() {
+                    chars.push_front(c)
+                }
```

Comment:
Would this add the characters in the reverse order?

For example, if `a` maps to `ABC`, will chars become:
```
['A']
['B', 'A']
['C', 'B', 'A']
```

Reply:
missed that, fixed.